broadinstitute/single_cell_portal_core
# Prevent clobbering differential expression output files (SCP-5222)

#### BACKGROUND & CHANGES
This update moves the cleanup of SCP-calculated differential expression results to before the ingest of author-provided results, instead of after successful parsing.  This prevents the output file cleanup from erroneously removing newly-generated results that may have had the same name as previous results, which just clobbered the existing auto-calculated files.  This scenario can happen if a user selects a cluster/annotation combination that already had results, and they use `wilcoxon` as the computational method.  If the uploaded DE file contains any one-vs-rest comparisons, then these files will be deleted if results already existed for those annotation values.

Additionally, if a study has author-provided DE results, this study will no longer be eligible for automatic differential expression calculations.  This is for two reasons: 1) we want to favor author-provided results, and 2) automatic calculation could potentially clobber existing author result files for the same reason as above.

#### MANUAL TESTING
Note: if you have visualization caching turned on, this can cause problems later in the Explore tab as the response is cached.  You can turn this off temporarily by running the following command from project root (ensuring you see the corresponding message):
```
bin/rails dev:cache
=> Development mode is no longer being cached.
```
1. Boot all services and sign in
2. Select any study that has DE results and go to the upload wizard
3. In a separate browser tab/window, open the Explore tab for the same study
4. Note that the `Differential expression` button is visible, meaning results are available
5. Back in the upload wizard, upload the [test author DE file](https://raw.githubusercontent.com/broadinstitute/scp-ingest-pipeline/development/tests/data/author_de/author_de_test_data_human_milk_All_Cells_UMAP_General_celltype.csv) and wait 10-15 seconds for the ingest process to launch
6. Back in the Explore tab, refresh the page and note that the `Differential expression` is gone
7. Open a Rails console session and confirm that this study is no longer eligible for automatic DE:
```
study = Study.find_by(accession: '<accession of study>')
DifferentialExpressionService.study_eligible?(study)
=> false
```
8. Optional: if you wish to restore the old DE results:
    i. Wait for the ingest to complete, then delete the test DE file
    ii. From the Rails console, run the following command:
```
DifferentialExpressionService.run_differential_expression_on_all("<accession of study>")
```
